### PR TITLE
Entity and component serialization

### PR DESCRIPTION
**Pull request by TheoBerlin**
The layout of the serializations are documented in `ECSCore.h`

### CODE REVIEW

File: LambdaEngine/Include/ECS/ComponentArray.h [hunk 1]
```diff
@@ -134,6 +141,59 @@ namespace LambdaEngine
...
+	template<typename Comp>
+	inline uint32 ComponentArray<Comp>::SerializeComponent(const Comp& component, uint8* pBuffer, uint32 bufferSize) const
+	{
+		/*	ComponentSerializationHeader is written to the beginning of the buffer. This is done last, when the size of
+			the serialization is known. */
+		struct ComponentSerializationHeader
+		{
+			uint32 TotalSerializationSize;
+			uint32 TypeHash;
+		};
```
Comment (TheRealChriAnd) [inline LambdaEngine/Include/ECS/ComponentArray.h:154]: To make sure no padding is added by the compiler in the future if adding more data to the header

```suggestion
        #pragma pack(push, 1)
		struct ComponentSerializationHeader
		{
			uint32 TotalSerializationSize;
			uint32 TypeHash;
		};
        #pragma pack(pop)
```

File: LambdaEngine/Include/ECS/ECSCore.h [hunk 2]
```diff
@@ -13,6 +13,14 @@ namespace LambdaEngine
 	class EntitySubscriber;
 	class RegularWorker;
 
+	// EntitySerializationHeader is written to the beginning of an entity serialization
+	struct EntitySerializationHeader
+	{
+		uint32 TotalSerializationSize;
+		Entity Entity;
+		uint32 ComponentCount;
+	};
```
Comment (TheRealChriAnd) [inline LambdaEngine/Include/ECS/ECSCore.h:23]: Same as the last comment
```suggestion
#pragma pack(push, 1)
	struct EntitySerializationHeader
	{
		uint32 TotalSerializationSize;
		Entity Entity;
		uint32 ComponentCount;
	};
#pragma pack(pop)	
```

File: LambdaEngine/Source/Networking/API/BinaryEncoder.cpp [hunk 3]
```diff
@@ -76,30 +76,24 @@ namespace LambdaEngine
...
+		WriteBuffer((const uint8*)&value.data, sizeof(glm::vec3));
 	}
 
 	void BinaryEncoder::WriteVec4(const glm::vec4& value)
 	{
-		WriteFloat32(value.x);
-		WriteFloat32(value.y);
-		WriteFloat32(value.z);
-		WriteFloat32(value.w);
+		WriteBuffer((const uint8*)&value.data, sizeof(glm::vec4));
```
Comment (TheRealChriAnd) [inline LambdaEngine/Source/Networking/API/BinaryEncoder.cpp:97]: We should update the BinaryDecoder to read from the data pointer as well

**Review by TheRealChriAnd — CHANGES_REQUESTED** [review]
Looking great! Please check the comments. We also need a way to deserialize the Entities, I guess that's for another task and PR.

**Comment by TheoBerlin** [comment]
> Looking great! Please check the comments. We also need a way to deserialize the Entities, I guess that's for another task and PR.

Applied all suggestions, please check again.

**Review by TheRealChriAnd — APPROVED** [review]
Nice!

**Pull request merged**